### PR TITLE
vscode web extension

### PR DESCRIPTION
**Pull request by ogoffart**
- builds the lsp in wasm  (this implies building as a library for wasm)
 - the wasm lsp cannot use the `lsp-server` crate to communicate because that uses stdin/stdout, and we need to communicate with the js runtime. So that's why the code there is a bit different.
 - The preview is not yet implemented
 - Currently only file that are open in the editor can be imported

### CODE REVIEW

File: editor/vscode/browser-language-server/browserServerMain.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        return true;
+    }
+
+    connection.onInitialize((params: InitializeParams): InitializeResult => {
+        the_lsp = slint_lsp.create(params, send_notification);
+        return { capabilities: the_lsp.capabilities() };
+    });
+
+    connection.onRequest((method, params, token) => {
+        return the_lsp.handle_request(token, method, params);
```
Comment (tronical) [inline editor/vscode/browser-language-server/browserServerMain.ts:30]: Ah, here the `token` becomes the `id`. Should it be used?

**Review by tronical — APPROVED** [review]
Terrific, can't wait to try it. I can't see any harm in merging this already.

File: editor/vscode/src/browser.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    };
+
+    const serverMain = Uri.joinPath(context.extensionUri, 'out/browserServerMain.js');
+    const worker = new Worker(serverMain.toString(true));
+    worker.onmessage = m => {
+        // We cannot start sending messages to the client before we start listening which
+        // the server only does in a future after the wasm is loaded.
+        if (m.data === "OK") {
+
+            client = new LanguageClient('slint-lsp', 'Slint LSP', clientOptions, worker);
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply editor/vscode/src/browser.ts:33]: yes.

File: tools/lsp/wasm_main.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            &ServerNotifier,
+            &mut self.0.borrow_mut(),
+        )
+        .map_err(|e| JsError::new(&e.to_string()));
+    }*/
+
+    #[wasm_bindgen]
+    pub fn handle_request(
+        &self,
+        _id: JsValue,
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply tools/lsp/wasm_main.rs:143]: The id is indeed the token that can be used if we want to cancel the request in the future.

but this is not used right now.

File: editor/vscode/package.json [hunk 4]
```diff
@@ -88,6 +89,7 @@
 	"scripts": {
 		"vscode:prepublish": "npm run compile",
```
Comment (tronical) [inline editor/vscode/package.json]: Maybe this helps with the CI:

```suggestion
		"vscode:prepublish": "npm run compile && npm run compile-web",
```

**Pull request merged**